File: content/snippets/python/interview_lcci_16_10_living_people.py
from typing import List


def max_alive_year(birth: List[int], death: List[int]) -> int:
    diff = [0] * 102
    for b, d in zip(birth, death):
        diff[b - 1900] += 1
        if d + 1 <= 2000:
            diff[d + 1 - 1900] -= 1

    best_year = 1900
    alive = 0
    best_alive = -1
    for i in range(101):
        alive += diff[i]
        if alive > best_alive:
            best_alive = alive
            best_year = 1900 + i
    return best_year


def main() -> None:
    print(max_alive_year([1900, 1901, 1950], [1948, 1951, 2000]))


if __name__ == "__main__":
    main()

File: content/snippets/typescript/interview_lcci_16_10_living_people.ts
function maxAliveYear(birth: number[], death: number[]): number {
    const diff = new Array<number>(102).fill(0);
    for (let i = 0; i < birth.length; i++) {
        diff[birth[i] - 1900] += 1;
        if (death[i] + 1 <= 2000) {
            diff[death[i] + 1 - 1900] -= 1;
        }
    }

    let bestYear = 1900;
    let alive = 0;
    let bestAlive = -1;
    for (let i = 0; i <= 100; i++) {
        alive += diff[i];
        if (alive > bestAlive) {
            bestAlive = alive;
            bestYear = 1900 + i;
        }
    }
    return bestYear;
}

console.log(maxAliveYear([1900, 1901, 1950], [1948, 1951, 2000]));
